### PR TITLE
Support error-binding-pattern as match-stmt

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
$subject 

Fixes ##26883

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java [hunk 1]
```diff
@@ -4252,6 +4258,15 @@ private BLangBindingPattern transformBindingPattern(Node bindingPattern) {
                             (BLangRestBindingPattern) transformBindingPattern(listMemberBindingPattern);
                 }
                 return bLangListBindingPattern;
+            case NAMED_ARG_BINDING_PATTERN:
```
Comment (Kishanthan) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:4261]: Lets move switch case logic that are too big to a separate method

Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:4263]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/bindingpatterns/BLangErrorCauseBindingPattern.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    public void accept(BLangNodeVisitor visitor) {
+        visitor.visit(this);
+    }
+
+    @Override
+    public NodeKind getKind() {
+
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/bindingpatterns/BLangErrorCauseBindingPattern.java:63]: Extra new line.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 3]
```diff
@@ -2814,14 +2820,31 @@ public void visit(BLangVarBindingPatternMatchPattern varBindingPattern) {
...
     }
 
     @Override
     public void visit(BLangWildCardBindingPattern wildCardBindingPattern) {
-        wildCardBindingPattern.type = symTable.anyType;
+        if (wildCardBindingPattern.type == null) {
+            wildCardBindingPattern.type = symTable.anyType;
+            return;
+        }
+        BType bindingPatternType = wildCardBindingPattern.type;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2841]: Can move this as the first line and change L2837 to use `bindingPatternType`.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 4]
```diff
@@ -2868,6 +2891,81 @@ public void visit(BLangRestBindingPattern restBindingPattern) {
...
 
+    @Override
+    public void visit(BLangErrorBindingPattern errorBindingPattern) {
+        if (errorBindingPattern.errorTypeReference != null) {
+            errorBindingPattern.type = symResolver.resolveTypeNode(errorBindingPattern.errorTypeReference, env);
+        } else {
+            errorBindingPattern.type = symTable.errorType;
+        }
+
+        if (errorBindingPattern.errorMessageBindingPattern != null) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2902]: Can extract `errorBindingPattern.errorMessageBindingPattern` out to a variable. Applies to the cause and field patterns too.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 5]
```diff
@@ -2868,6 +2891,81 @@ public void visit(BLangRestBindingPattern restBindingPattern) {
...
+        }
+    }
+
+    @Override
+    public void visit(BLangSimpleBindingPattern simpleBindingPattern) {
+        if (simpleBindingPattern.wildCardBindingPattern != null) {
+            analyzeNode(simpleBindingPattern.wildCardBindingPattern, env);
+            return;
+        }
+        if (simpleBindingPattern.captureBindingPattern != null) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2924]: Can this be null if `simpleBindingPattern.wildCardBindingPattern` is null?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 6]
```diff
@@ -2868,6 +2891,81 @@ public void visit(BLangRestBindingPattern restBindingPattern) {
...
+    }
+
+    @Override
+    public void visit(BLangErrorCauseBindingPattern errorCauseBindingPattern) {
+        if (errorCauseBindingPattern.simpleBindingPattern != null) {
+            analyzeNode(errorCauseBindingPattern.simpleBindingPattern, env);
+            errorCauseBindingPattern.declaredVars.putAll(errorCauseBindingPattern.simpleBindingPattern.declaredVars);
+            return;
+        }
+        if (errorCauseBindingPattern.errorBindingPattern != null) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2944]: Can this be null if `errorCauseBindingPattern.simpleBindingPattern` is null?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java [hunk 7]
```diff
@@ -4123,59 +4129,88 @@ private BLangMatchPattern transformMatchPattern(Node matchPattern) {
...
-            createErrorFieldMatchPatterns(2, errorMatchPatternNode, bLangErrorMatchPattern);
+        if (kind == SyntaxKind.MAPPING_MATCH_PATTERN) {
+            return transformMappingMatchPattern((MappingMatchPatternNode) matchPattern, matchPatternPos);
+        }
+
+        if (kind == SyntaxKind.FIELD_MATCH_PATTERN) {
+            return transformFieldMatchPattern((FieldMatchPatternNode) matchPattern, matchPatternPos);
+        }
+
+        // TODO : Remove this after all binding patterns are implemented
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:4155]: Can we remove this now?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 8]
```diff
@@ -3682,6 +3693,133 @@ private BLangExpression createConditionForMappingBindingPattern(BLangMappingBind
         return statementExpression;
     }
 
+    private BLangExpression createConditionForErrorBindingPattern(BLangErrorBindingPattern errorBindingPattern,
+                                                                  BLangSimpleVarRef matchExprVarRef) {
+        BType bindingPatternType = errorBindingPattern.type;
+        Location pos = errorBindingPattern.pos;
+
+        BLangSimpleVariableDef resultVarDef = createVarDef("errorBindingPatternResult$", symTable.booleanType, null,
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:3701]: Same comment as the previous PR regarding using a counter and moving to `BLangAnonymousModelHelper`.

**Review by MaryamZi — APPROVED** [review]
@KavinduZoysa, please create an issue to track fixing the suggestions.

**Pull request merged**